### ISSUE TITLE
C# client library fails on TLS authentication error on downstream device  while Python works

### ISSUE THREAD

**Issue by ondrejspilka** [Azure/iotedge#833]
## Expected Behavior
C# sample provided for downstream IoT edge devices shall work in the same way as Python example.
Similar to #209 

## Resolution
See edit part downbelow

## Current Behavior
C# sample ends with TLS authentication error `The remote certificate is invalid according to the validation procedure`

## Steps to Reproduce
1. Use python example on the same machine - works with no issues
2. Use C# example written in .NET Core, using same connection string, above error aplies

## Context (Environment)
Downstream device running Windows server

### Device (Host) Operating System
RHEL 7.5

### Architecture
x64

### Container Operating System
Linux containers

### Runtime Versions
IoT edge 1.0.6

## Logs
`Unhandled Exception: System.AggregateException: One or more errors occurred. (TLS authentication error.) ---> System.Security.Authentication.AuthenticationException: TLS authentication error. ---> System.AggregateException: One or more errors occurred. (The remote certificate is invalid according to the validation procedure.) ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Security.SslState.ThrowIfExceptional()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__46_2(IAsyncResult iar)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttIotHubAdapter.WriteMessageAsync(IChannelHandlerContext context, Object message, Func`3 exceptionHandler)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttIotHubAdapter.Connect(IChannelHandlerContext context)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass35_0.<<OpenAsyncInternal>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.EnsureOpenedAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass17_0.<<SendEventAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendEventAsync(Message message, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.SendEventAsync(Message message)
   at IoTTest.Net.Core.Program.SendEvents(DeviceClient deviceClient, Int32 messageCount) 
`

## Additional Information
Python sample works with no issues.
Certificates are installed in Computer store, we have tried Personal store as well.
Openssl check for generated PFX works well.
Device name is correct and equal to the one used in certificate generation procedure.
See my comments to Python sample in Product feedback on [https://docs.microsoft.com/en-us/azure/iot-edge/how-to-connect-downstream-device](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-connect-downstream-device) 

## Edit - Resolution
Took me some time, happily IoT hub C# SDK is available so after reading code a bit I understood.
The trick is that C# IoT MQTT ChannelFactory deduces hostname used for TLS settings from `iotHubConnectionString.HostName`
When certificate chain for development is issued, according to [documentation](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-create-transparent-gateway) when creating self-signed certificates, `"<gateway name>"` has to be chosen. This must be the same as the name passed to TLS connection settings, otherwise TLS will fail on hostname check.

The implication is that in contradiction to Python, where FQDN can be used for gateway host, in C# one has to:

- use FQDN for certificate name
- or do not use FQDN for Gateway host

So if certificates for gateway are issued with gateway name `MYIOTEDGEDEVICE`
then in Python connection string might look like:

`HostName=myIoTHub.azure-devices.net;
DeviceId=MYIOTEDGEDEVICE;
SharedAccessKey=mySAS;
GatewayHostName=myiotedgedevice.aa.bb.com
`

and it works, because in Python FQDN can be used as a gateway host while for TLS verification DeviceId is properly used.
In C# above doesn't work and following connection string has to be:

`HostName=myIoTHub.azure-devices.net;
DeviceId=MYIOTEDGEDEVICE;
SharedAccessKey=mySAS;
GatewayHostName=myiotedgedevice
`

IMHO this is bug in device client, which has to use DeviceId for gateway host certificate name in TLS.
Source is in [MqttTransportHandler](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs)

Finally, still I got some RemoteCertificateNameMismatch SSL errors, which, if ommited, let messages pass through gateway to IoT Hub:

``` csharp
var settings = new MqttTransportSettings(TransportType.Mqtt_Tcp_Only);
settings.Proxy = null;
settings.RemoteCertificateValidationCallback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
{
	Console.WriteLine("SSL Error {0}\n", sslPolicyErrors);
   return true;
};

DeviceClient deviceClient = DeviceClient.CreateFromConnectionString(DeviceConnectionString, new ITransportSettings[] { settings });

```

**Comment by myagley** [comment]
cc @mrohera

**Comment by mrohera** [comment]
Hello @ondrejspilka,
First off thanks for taking the time to go through the scenario and suggest feedback. This is really awesome to see. You have given me a lot of information here and I will have to try and repro your steps to see how to incorporate these. Please stay tuned for an update.

**Comment by tanieee28** [comment]
Hi,
I am also facing kind of similar error in my windows 10 machine. I am trying to deploy simulated temperature module --> mcr.microsoft.com/azureiotedge-simulated-temperature-sensor:1.0 to my machine but after container creation i am facing problems.

**

- Logs

**
[2019-02-21 12:23:02 : Starting Module
[02/21/2019 12:23:03.012 PM] Main()
Initializing simulated temperature sensor to send 500 messages, at an interval of 5 seconds.
To change this, set the environment variable MessageCount to the number of messages that should be sent (set it to -1 to send unlimited messages).
Using transport Amqp_Tcp_Only

Unhandled Exception: System.AggregateException: One or more errors occurred. (The remote certificate is invalid according to the validation procedure.) ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
   at Microsoft.Azure.Devices.Client.InternalClient.<>c.<ApplyTimeout>b__62_2(Task t)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at SimulatedTemperatureSensor.Program.InitModuleClient(TransportType transportType) in /home/vsts/work/1/s/edge-modules/SimulatedTemperatureSensor/src/Program.cs:line 114
   at SimulatedTemperatureSensor.Program.MainAsync() in /home/vsts/work/1/s/edge-modules/SimulatedTemperatureSensor/src/Program.cs:line 81
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at SimulatedTemperatureSensor.Program.Main() in /home/vsts/work/1/s/edge-modules/SimulatedTemperatureSensor/src/Program.cs:line 39



**Request somebody to get help on this.**

**Comment by myagley** [comment]
What is the `hostname` set to in your `C:\ProgramData\iotedge\config.yaml`?

**Comment by tanieee28** [comment]
Sorry for responding late.
Thanks for the reply.

hostname: 'Shakatiman_I'
above is the line present in config.yaml which represents hostname. 
I have also checked the OS properties, hostnames are same.

I have got to know that the problem is only with this machine, i tried to deploy it on other machine and the runtime got deployed successfully.
Checked all the possible permutations and combinations which can cause the error, still no success on this.

**Comment by myagley** [comment]
What version is the `iotedged` on this machine (run `iotedge version`)? There is a bug in `1.0.6` which will cause hostnames with an `_` to fail cert validation. A patch release `1.0.6.1` was released to fix this.

**Comment by tanieee28** [comment]
![iotedge_version_check](https://user-images.githubusercontent.com/22118756/53572680-cef57400-3b91-11e9-941e-4c4335ea658e.png)

Above is the screenshot of command : **"iotedge version"**

Yes, you are right, when i googled this error i got similar comments for problem in hostname with **'_'** special character. But looks like i have the updated version.

**Comment by myagley** [comment]
This was fixed in 1.0.6.1. Closing due to inactivity.
